source/common/http/utility.cc

```diff
@@ -463,13 +466,22 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
       }
       response_headers->setGrpcMessage(PercentEncoding::encode(body_text));
     }
+    // The `modify_headers` function may have added content-length, remove it.
+    response_headers->removeContentLength();
     encode_functions.encode_headers_(std::move(response_headers), true); // Trailers only response
     return;
   }
 
   if (!body_text.empty()) {
     response_headers->setContentLength(body_text.size());
-    response_headers->setReferenceContentType(content_type);
+    // If the `rewrite` function has changed body_text or content-type is not set, set it.
+    if (body_text != local_reply_data.body_text_ || response_headers->ContentType() == nullptr) {
+      response_headers->setReferenceContentType(content_type);
+    }
+  } else {
+    // The `modify_headers` function may have added content-length and content-type, remove them.
+    response_headers->removeContentLength();
```

Comment:
Good catch on remocal, but in the case of HTTP/1.1 having an explicit content length of 0 is useful, so maybe just force update the content length to the length of body_text if it's empty or not?

Reply:
Set Content-Length to 0 for empty body break some tests.  Decided not to add or remove any headers for empty body. Try to minimize the changes as much as possible.

Reply:
I think without those changes the code paths might still be buggy if the body is removed. cc @mattklein123 in case he has thought on this.

Reply:
fwiw I'm fine with skipping content-length:0 - my main goal was to avoid frame by connection close and as long as we encode headers with end_stream=true the HTTP/1 codec takes care of that.  I think removing invalid content lengths is still worth doing though.

Reply:
Sorry not fully tracking the convo here. I agree that we should be able to set content-length in all cases, but I would be OK with minimizing changes for now.

Reply:
OK, I restored the code back to the previous version.  
When body is empty, remove ContentType and ContentLength headers.

Also add an new DirectResonse Integration test to test the new logic.